### PR TITLE
HW4: Patterns

### PR DESCRIPTION
**Pull request by CalculusEnjoyer**
* Add Repository pattern in Storage microservice (services/storage/emails/storage_service.go) and in Api microservice that uses Storage as database basically (services/api/rest/ctrl/email_controller.go) 
* Add 2 new rate providers and implement the Chain of Responsibility pattern in order to use the next provider if the previous one fails. Bootstrapping all chain in `bootstrapRateProviders()`  (services/currency/main.go)
* Add logging in the console what provider was used.

Additionally: edited docker-compose, now passing env variables through it and remove unnecessary port exposes

### CODE REVIEW

**Review by Miha-s — COMMENTED** [review]
Generally everything looks quite well ✨
Especially I liked RateChain concept

But I would say that there are some code duplication
Each your providers has method currencyToString, which is identical except map with currencies convertion. Maybe it would be better to make some global function with that logic, and pass individual currencies maps to that function, etc.

Also there are some inconsistency in code of Providers. CoinGeckoRateProvider check for decode error in GetExchangeRate and logs the rate. Others do that in decodeRateResponse.
If you generalize that a bit, you'll see, that there are some almost identical patches of code, and maybe you will find a solution to eliminate those duplication.

Cheers, Mike

**Comment by CalculusEnjoyer** [comment]
> Generally everything looks quite well ✨ Especially I liked RateChain concept
> 
> But I would say that there are some code duplication Each your providers has method decodeRateResponse(response, convertedBase, convertedTarget), which is identical except map with currencies convertion. Maybe it would be better to make some global function with that logic, and pass individual currencies maps to that function, etc.
> 
> Also there are some inconsistency in code of Providers. CoinGeckoRateProvider check for decode error in GetExchangeRate and logs the rate. Others do that in decodeRateResponse. If you generalize that a bit, you'll see, that there are some almost identical patches of code, and maybe you will find a solution to eliminate those duplication.
> 
> Cheers, Mike

Thanks for the review, Mike😎!

Yeah, `currencyToString` can be generalized, good point.
According to the generalizing logic of decoding rate request, IMHO it does not worth it and saves only a couple of lines of code, but dramatically decreases code readability

File: services/currency/rate/providers/crypto/coingecko_provider.go [hunk 1]
```diff
@@ -3,39 +3,74 @@ package crypto
...
 
-	rates := data[strings.ToLower(baseCurrency)]
-	rate = rates[strings.ToLower(targetCurrency)]
+	rates := data[baseCurrencyName]
+	rate, ok := rates[targetCurrencyName]
+	if !ok {
+		return cerror.ErrRateValue, cerror.ErrRate
+	}
 
+	log.Printf("Getting rate from COINGECKO: %f", rate)
```
Comment (kr0zer0) [inline services/currency/rate/providers/crypto/coingecko_provider.go:66]: логування в даному випадку варто було додавати за допомогою декоратора. тоді не доведеться дублювати і легко можжна буде вмикати/вимикати. Щоб передати  в лог назву провайдера можна додати в абстракцію провайдера щось на кшталт Name()

**Comment by kr0zer0** [comment]
не можу знайти кешування відповіді від провайдера, ти його додавав?

File: services/currency/rate/rate_chain.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
+package rate
+
+import (
+	"currency/rate/messages"
+)
+
+type RateLink struct {
+	next     *RateLink
+	provider RateProvider
+}
```
Comment (kr0zer0) [inline services/currency/rate/rate_chain.go:10]: з Chain of responsibility све круто, молодець. Єдине чого не вичтачає - абстракції для RateLink, варто додати ще інтерфейс та використовувати його

File: services/currency/rate/providers/crypto/coingecko_provider.go (hunk 1 above)
Comment (CalculusEnjoyer) [reply services/currency/rate/providers/crypto/coingecko_provider.go]: справді, має сенс, переробив використовуючи декоратор 👍

**Comment by CalculusEnjoyer** [comment]
> не можу знайти кешування відповіді від провайдера, ти його додавав? Також не помітив facory method

@kr0zer0 , та нам здається не казали поки робити кешування відповіді, то я й не робив 😀. Але так, резонно, доробив в новому коміті кешування використовуючи проксі 👍

Щодо factory method, використовую тут просто creation method, бо здалося що так буде простіше і читабельніше і що у factory method тут поки немає потреби.

**Pull request merged**